design/idempotency_ids.md

```diff
@@ -0,0 +1,103 @@
+# Goals
+
+The main goal is to make transactions safer and easier to reason about. New users should get a "just works" experience. One of the main selling points of foundationdb is that it solves the hard distributed systems problems for you, so that you only need to concern yourself with your business logic. Non-idempotent transactions is probably the biggest "gotcha" that users need to be made aware of - and they won't discover it organically. In order to achieve this "just works" experience I believe it is necessary to make automatic idempotency have low-enough overhead that we can enable it by default.
+
+As an intermediate goal, I plan to introduce this feature disabled by default. The long-term plan is to make it the default.
+
+# API
+
+Introduce a new transaction option `IDEMPOTENCY_ID`, which will be validated to be at most 255 bytes.
+Add 
+```
+FDBFuture* fdb_transaction_commit_result(FDBTransaction* tr, uint8_t const* idempotency_id, int idempotency_id_length)
+```
+, which can be used to determine the result of a commit that failed with `transaction_timed_out`.
+
+Commits for transactions with idempotency ids would not fail with `commit_unknown_result`, but in (extremely) rare cases could fail with a new error that clients are expected to handle by restarting the process.
+# Background
+
+- https://forums.foundationdb.org/t/automatically-providing-transaction-idempotency/1873
+- https://github.com/apple/foundationdb/issues/1321
+- https://docs.google.com/document/d/19LDQuurg4Tt8eUcig3-8g2VOG9ZpQvtWrp_691RqMo8/edit#
+
+# Data model
+
+Commit proxies would combine idempotency ids for transactions within a batch. The purpose of this is to try to limit the number of distinct database keys that need to be written, and to lessen the number of extra mutation bytes for idempotency ids.
+
+## Key format
+```
+\xff\x02/idmp/${commit_version_big_endian (8 bytes)}${high_order_byte_of_batch_index (1 byte)}
+```
+
+- `commit_version_big_endian` the commit version stored big-endian so that the cleaner worker can find the oldest idempotency ids easily, and also so that "unknown_committed" transactions can recover their commit version.
+- `high_order_byte_of_batch_index` this limits us to 256 idempotency ids per value
+
+## Value format
+```
+${protocol_version}(${n (1 byte)}${idempotency_id (n bytes)}${low_order_byte_of_batch_index})*
+```
+
+The batch index for each idempotency id can be reconstructed from the high order byte and low order bytes stored in the key and value, respectively. This is necessary for an "unknown_committed" transaction to recover their full version stamp.
```

Comment:
Maybe explicitly mention that batch index is `short int`, i.e., 2 bytes.

Reply:
Done (not pushed yet)